src/utils/aadV2TokenProvider.ts

```diff
@@ -0,0 +1,264 @@
+                value = await window.showInformationMessage(donePrompt, messageBoxOptions, done, tryAgain);
+            } while (value === tryAgain);
+        }
+        return value === done;
+    }
+}
+
+/*
+
+  ClientId: We use default clientId for all delegated access unless overridden in $appToken.  AppOnly access uses the one in the environment
+  TenantId: If not specified, we use common. If specified in environment, we use that. Value in $aadToken overrides
+  Scopes are always in $aadV2Token for delegated access. They are not used for appOnly.
+*/
+class AuthParameters {
+
+    private readonly aadV2TokenRegex: RegExp = new RegExp(`\\s*\\${Constants.AzureActiveDirectoryV2TokenVariableName}(\\s+(${Constants.AzureActiveDirectoryForceNewOption}))?(\\s+(appOnly))?(\\s+scopes:([\\w,.]+))?(\\s+tenantId:([^\\.]+\\.[^\\}\\s]+|[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}))?(\\s+clientId:([^\\.]+\\.[^\\}\\s]+|[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}))?\\s*`);
+    public tenantId: string;
+    public clientId: string;
+    public scopes: string[];
+    public forceNewToken: boolean;
+    public clientSecret?: string;
+    public appOnly: boolean;
+    public appUri?: string;
+
+    public constructor() {
+        this.clientId = AadV2TokenProviderClientId;
+        this.tenantId = "common";
+        this.forceNewToken = false;
+        this.appOnly = false;
+    }
+
+    async readEnvironmentVariable(variableName: string) : Promise<string | undefined> {
+        if (await EnvironmentVariableProvider.Instance.has(variableName)) {
+            const { value, error, warning } = await EnvironmentVariableProvider.Instance.get(variableName);
+            if (!warning && !error) {
+                return value as string;
+            } else {
+                return undefined;
+            }
+        }
+        return undefined;
+    }
+
+    getCacheKey() : string {
+        return this.tenantId + "|" + this.clientId + "|" + this.appOnly as string;
+    }
+
+    async parseName(name: string) {
+
+        // Update defaults based on environment
+        this.tenantId = (await this.readEnvironmentVariable("aadV2TenantId")) || this.tenantId;
+
+        let scopes = "openid,profile";
+        let explicitClientId: string | undefined = undefined;
+        // Parse variable parameters
+        const groups = this.aadV2TokenRegex.exec(name);
+        if (groups) {
+            this.forceNewToken = groups[2] === Constants.AzureActiveDirectoryForceNewOption;
+            this.appOnly = groups[4] === "appOnly";
+            scopes = groups[6] || scopes;
+            this.tenantId = groups[8] || this.tenantId;
+            explicitClientId = groups[10];
+        } else {
+            throw new Error("Failed to parse parameters: " + name);
+        }
+
+        this.scopes = (scopes + ",openid,profile").split(",").map(s => s.trim());
```

Comment:
It seems that `this.scopes` may contain multiple `profile`

Reply:
Fixed and added a comment to why we are adding these scopes.